capstone/src/main/java/com/google/sps/servlets/GroupPostDataServlet.java

```diff
@@ -0,0 +1,94 @@
+package com.google.sps.servlets;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.HashSet;
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import com.google.sps.Objects.User;
+import com.google.sps.Objects.Post;
+import com.google.sps.Objects.Comment;
+
+@WebServlet("/group-post")
+public class GroupPostDataServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    Query query = new Query("Post").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Post> posts = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      posts.add(getPostEntity(entity));
+    }
+
+    // Convert to json
+    response.setContentType("application/json;");
+    response.getWriter().println(new Gson().toJson(posts));
+  }
+
+  public Post getPostEntity(Entity entity) {
+    long timestamp = (long) entity.getProperty("timestamp");
+    String authorId = (String) entity.getProperty("authorId");
+    String postText = (String) entity.getProperty("postText");
+    String challengeName = (String) entity.getProperty("challengeName");
+    String img = (String) entity.getProperty("img");
+    List<String> likes = (ArrayList<String>) entity.getProperty("likes");
+    HashSet<String> likesSet = new HashSet<String>(likes);
```

Comment:
I apologize since I know I told you to do this, but this is the last change and then I'll approve. Since likes isn't being used anywhere, can you inline it when you create the HashSet and then change the name of the HashSet to likes? Ideally the variable names shouldn't need to have the datatype baked into the name.
```suggestion
    HashSet<String> likes = new HashSet<String>((ArrayList<String>) entity.getProperty("likes"));
```

Reply:
No worries! I change it to an in-line statement. I was getting a null pointer exception so I added an additional check.